### ISSUE TITLE
Wait for SSH Connectivity before local-exec (Ansible)

### ISSUE THREAD

**Issue by IanVS** [hashicorp/terraform#2811]
I am using Ansible for provisioning by kicking off `ansible-playbook ...` as a local-exec script.  

Similar to https://github.com/hashicorp/terraform/issues/2137, I find that the local-exec is executed as soon as an EC2 instance is created, but before SSH connectivity is established, and SSH times out.

What is the right way to handle this case?  Do I just need to put a long ssh timeout in Ansible?  That seems brittle.  Is it not possible to delay execution of `local-exec` provisioners until SSH is ready?  There must be some kind of logic for that already to allow `remote-exec` scripts to work.

**Comment by mlrobinson** [comment]
I don't know of a terraform way of solving this problem, but there is an Ansible module to help you with it:

http://docs.ansible.com/ansible/wait_for_module.html

That (ran as a local action) would allow ansible to start immediately, but then wait until you saw the port become available. Not a perfect solution, but it should work for you for now.

**Comment by IanVS** [comment]
Thanks, I'll give that a shot.  Still would be a little more elegant if there was a way to handle in Terraform...

**Comment by IanVS** [comment]
Actually I have a question about this, since you seem to know what you're talking about, @mlrobinson.  I'm using roles to set up my boxes, and my top level playbook is empty except for specifying a role.  So, would this `wait_for` task need to go into my `common` role which is the base dependency of my roles?  I think if I put it in the playbook, it would be executed after all the roles are set up, correct?  Thanks again for the advice.

**Comment by mlrobinson** [comment]
That depends on your role setup of course. If all your roles include the `common` role first, then it should run before everything else. Another option would be to make this it's own, very simple role, and just define that before even common. Either option would be viable, but I'd probably go with the easier of the two, depending on how your playbooks/roles are setup.

But, according to http://docs.ansible.com/ansible/playbooks_roles.html you are right that a "loose" task in the same playbook as roles is executed after the roles have been finished. Someone online created a "command role" that basically allowed you to pass a string into the role definition so it would occur in order, but still support defining one off commands at the playbook level. Here's one example of it: https://github.com/RoleModel/bash_command

**Comment by mlrobinson** [comment]
Not to double post, but ignore me about the command role. That won't work exactly in your situation, though something like it is what I was thinking of: a role that allowed you to pass in the name of another module to run. Such a meta thing would probably not be worth it in the run, but it would be a neat solution to this and other things with Ansible.

**Comment by IanVS** [comment]
Nope, this doesn't work.  Ansible times out at the `"Gathering Facts"` stage, even before any tasks get a chance to execute.  My only solution is to set the `timeout` variable in `ansible.cfg` to something like 10 minutes to be safe.

**Comment by mlrobinson** [comment]
Hmm. You could set "gather_facts" to false, but I'd imagine you need that for other things to work properly. Sorry it didn't work.

**Comment by MattSurabian** [comment]
I've just started using Terraform with Ansible via atlas and from what I've seen the AWS provider is waiting for the servers to enter a ready state before kicking off a local-action. Maybe this has been fixed somewhere along the line already?

**Comment by phinze** [comment]
There's not much fancy on the Terraform side of this. Internally, the SSH communicator used by the `remote-exec` provisioner just loops until SSH is available.

It's true that in a recent release we switched the status we wait to see on the AWS side before we consider the instance started, but you still inevitably have to wait for the machine to boot and launch SSH.

I could picture us adding an option to the local-exec provisioner like `wait_for_ssh`, but that gets sort of implementation specific (e.g. what about WinRM?). So part of me feels like it's easiest to just say "poll for SSH in your script if you need it". 

I'd welcome further thoughts from others on this thread though. :thought_balloon:

**Comment by amitsehgal** [comment]
Is this issue resolved ? This is big for me too...maintaning wait_for in ansible for everything is ugly

**Comment by phinze** [comment]
Just reviewed this thread and it looks like everybody asking for this behavior is using Ansible. If we added an Ansible-specific provider, we could make this waiting behavior baked in (#2661). Would that address the need enough that we can avoid adding communicator attributes to local-exec?

**Comment by MattSurabian** [comment]
I would love to see an Ansible specific provider, but that's probably old news :-p

I did finally run into this issue even though in my initial testing it wasn't happening. What I've ended up doing on complex deployments to get around this problem is use a local Ansible playbook that calls Terraform via the `command` module then runs a `wait_for` task that makes sure SSH is available on the DNS records of the newly created boxes before continuing by making `command` calls to other playbooks for the actual provisioning.

I'd love to not have to do all that and run everything through Terraform instead.

**Comment by djallits** [comment]
+1 to Ansible specific provider!

**Comment by jonmorehouse** [comment]
Hey guys! I've been working on an `ansible` provider here. Can follow along at https://github.com/jonmorehouse/terraform-provisioner-ansible

**Comment by booch** [comment]
As a work-around, perhaps we could just use a `remote-exec` provisioner (that just does an `ls` or something) and then run Ansible via the `local-exec` provisioner after that.

**Comment by panda87** [comment]
@booch Did you try to run more then one playbooks on one instance? if so, does Terraform support multiple commands in one local-exec or needs to run provisioner for each playbook?

**Comment by booch** [comment]
@panda87 I haven't tried using more than 1 playbook or instance yet. I think through Terraform, you'd need to have both the `remote-exec` and `local-exec` sections work for a single host. But when using Ansible after you've provisioned the servers, you'd want Ansible to work for multiple hosts.

I know that you can run multiple commands in a `remote-exec` section. Here's the one I use after running Ansible:

~~~
    provisioner "remote-exec" {
      # Can't run these via root Ansible, as they'd take away our SSH access.
      inline = [
          # Don't allow SSH access directly to root.
          "rm /root/.ssh/authorized_keys",
          "sudo sed -i -e 's/^PermitRootLogin .*$/PermitRootLogin forced-commands-only/' /etc/ssh/sshd_config",
      ]
    }
~~~

**Comment by vmindru** [comment]
I wonder why this one is still open.  the correct solution would be to use a remote-exec right before local-exec with ansible, remote-exec will retry in a loop until SSH is ready. 

VM

**Comment by jangrewe** [comment]
My solution:
```
    provisioner "remote-exec" {
        script = "scripts/wait_for_instance.sh"
    }

    provisioner "local-exec" {
        command = "ansible-playbook ..."
    }
```

and `wait_for_instance.sh` contains:
```
#!/bin/bash

while [ ! -f /var/lib/cloud/instance/boot-finished ]; do
  echo -e "\033[1;36mWaiting for cloud-init..."
  sleep 1
done
```

This makes sure that the instance has completely finished booting and initialization.

**Comment by neurogenesis** [comment]
If you use a bastion host, you'll get multiple output lines for each failed attempt to reach the instance (before the resource is available).

```
...
aws_instance.web_01: Creating...
aws_instance.web_01: Still creating... [10s elapsed]
aws_instance.web_01: Still creating... [20s elapsed]
aws_instance.web_01: Still creating... [30s elapsed]
aws_instance.web_01: Provisioning with 'remote-exec'...
aws_instance.web_01 (remote-exec): Connecting to remote host via SSH...
aws_instance.web_01 (remote-exec):   Host: x.x.x.x
aws_instance.web_01 (remote-exec):   User: ~/.ssh/bootstrap.pem
aws_instance.web_01 (remote-exec):   Password: false
aws_instance.web_01 (remote-exec):   Private key: true
aws_instance.web_01 (remote-exec):   Certificate: false
aws_instance.web_01 (remote-exec):   SSH Agent: true
aws_instance.web_01 (remote-exec):   Checking Host Key: false
aws_instance.web_01 (remote-exec): Using configured bastion host...
aws_instance.web_01 (remote-exec):   Host: bastion
aws_instance.web_01 (remote-exec):   User: xxx
aws_instance.web_01 (remote-exec):   Password: false
aws_instance.web_01 (remote-exec):   Private key: true
aws_instance.web_01 (remote-exec):   SSH Agent: true
aws_instance.web_01 (remote-exec):   Checking Host Key: false
aws_instance.web_01 (remote-exec): Connecting to remote host via SSH...
aws_instance.web_01 (remote-exec):   Host: x.x.x.x
aws_instance.web_01 (remote-exec):   User: ~/.ssh/bootstrap.pem
aws_instance.web_01 (remote-exec):   Password: false
aws_instance.web_01 (remote-exec):   Private key: true
aws_instance.web_01 (remote-exec):   Certificate: false
aws_instance.web_01 (remote-exec):   SSH Agent: true
aws_instance.web_01 (remote-exec):   Checking Host Key: false
aws_instance.web_01 (remote-exec): Using configured bastion host...
aws_instance.web_01 (remote-exec):   Host: bastion
aws_instance.web_01 (remote-exec):   User: xxx
aws_instance.web_01 (remote-exec):   Password: false
aws_instance.web_01 (remote-exec):   Private key: true
aws_instance.web_01 (remote-exec):   SSH Agent: true
aws_instance.web_01 (remote-exec):   Checking Host Key: false
aws_instance.web_01: Still creating... [40s elapsed]
...
```

Potentially a couple hundred lines per server.

I can see a few different types of `wait_for` or `depends_on` conditions that would be super handy. Some examples:

- waiting for the server or resource to become available (reduce unnecessary connection attempts or failed login attempts on bastion hosts)
- waiting for a local null resource, like presence of record in registration database or CMDB
- waiting for a callback / webhook (via lambda & SNS, or other provider equivalent)

This would cut down on unnecessary output for automated provisioning, it would also reduce failed login attempts and potential blacklisting on bastion hosts.

**Comment by neurogenesis** [comment]
@phinze : ^^  operations that can't run until a resource is available probably shouldn't attempt to do so. I would understand looping ssh calls after a resource has been created, but doing so beforehand doesn't make much sense.

**Comment by teamterraform** [comment]
Hello,

The ansible-specific nature of this issue was solved by the ansible provider, so I think we can close this out now.

As for waiting for connectivity before ssh-ing, the only way to check for ssh connectivity is to attempt a connection. There's no way for Terraform to be notified when a server starts an ssh daemon, or opens a specific port.  Since Terraform has no introspection into the server internals, it is often better to leverage existing Os-specific facilities for this, like cloud-init for example. 

The `remote-exec` provisioner is likely to remain a simple wrapper around ssh and a shell script for the time being. If there are specific concerns or feature requests for provisioners, they can be opened in a new issue.

Thanks
